### DISCUSSION TITLE
다음 코드의 출력 결과가 무엇인지 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#97] [this]
- 다음 코드의 출력 결과가 무엇인지 설명해주세요.

```js
function Person(name) {
  this.name = name;
}

var person = Person('John');
console.log(person);
console.log(person.name);

var person = new Person('John');
console.log(person);
console.log(person.name);
```

**Comment by publdaze** [comment]
> * 다음 코드의 출력 결과가 무엇인지 설명해주세요.
> 
> ```js
> function Person(name) {
>   this.name = name;
> }
> 
> var person = Person('John');
> console.log(person);
> console.log(person.name);
> 
> var person = new Person('John');
> console.log(person);
> console.log(person.name);
> ```

- 첫번째 person에 할당되는 값은 일반함수로서 호출된 함수의 반환 값
- 함수가 호출되면서 this.name에 name이 할당되는데, 이때 this는 전역객체로, 전역객체의 프로퍼티로 name 할당
- 첫번째 출력은 반환값이 따로 없어 undefined가 출력되고,
- 두번째 출력은 undefined에 name이라는 프로퍼티가 존재하지 않으므로 타입에러 발생

// 에러 발생 후 실행이 멈추지만 이어서 진행된다 가정하면

- 두번째 person에 할당되는 값은 생성자 함수의 인스턴스이므로
- this에는 인스턴스인 person가 바인딩 됨
- 첫번째 출력은 Person { name: 'John' } 출력
- 두번째 출력은 this가 인스턴스에 바인딩되어 있으므로 'John' 출력

**Comment by pipisebastian** [comment]
> * 다음 코드의 출력 결과가 무엇인지 설명해주세요.
> 
> ```js
> function Person(name) {
>   this.name = name;
> }
> 
> var person = Person('John');
> console.log(person);
> console.log(person.name);
> 
> var person = new Person('John');
> console.log(person);
> console.log(person.name);
> ```

- undefined, TypeError, Person {name : John}, John
- [1] 일반 함수로 호출했기 때문에, person에는 Person함수의 return값이 들어간다. 그러나 아무것도 return하지 않기 때문에, undefined가 출력된다
- [2] person은 undefined인데, person 의 name 속성을 참조하기때문에, TypeError를 발생한다
- [3] new로 새로운 인스턴스를 생성했기에, person 인스턴스를 출력한다. 
- [4] person 인스턴스의 name 속성, John 을 출력한다.
